README
LEOCHARRE::Dev
This distro is a set of tools to help me manage a distro- and development.
It is in my namespace because a lot of the scripts and methods here are already available in other packages.

I find these set of tools very valuable.

For example, 

   - automatically generate pod from source
   - analize your distro and tell what modules are deps and what versions are currently installed
   - auto Makefile.PL and MANIFEST generation
   - check out what version of a module is installed and if it's the most current on cpan
   - etc

To install

   perl Makefile.PL
   make install




Changes
2009-08-29 21:20  cvs

	* LEOCHARRE-Dev-1.03.tar.gz, Makefile.PL, README,
	  lib/LEOCHARRE/Dev.pm: Made sure proper license shows up, perl
	  gpl.	Cleaned up docs and README

2009-08-14 11:17  cvs

	* MANIFEST, Makefile.PL, bin/pmuseoff: removed pmuseoff, because we
	  have the more current and better documented bin/pmonoff

2009-08-14 11:10  cvs

	* bin/pmonoff: added pmonoff, to edit code in perl files, turning
	  module use off or on, optionally overriting also backs up old
	  file

2009-06-24 14:40  cvs

	* bin/pmreadme: updated so readme will contain current year

2009-05-18 16:55  cvs

	* bin/pmmakefile: made sure if .h or .h or pod files are in bin,
	  they are not treated as exe files

2009-05-05 15:01  cvs

	* bin/pmclassinspector: added conveninent cli class inspector

2009-04-09 15:49  cvs

	* bin/pmmanifest: made so manifest quotes if the filename has
	  spaces

2009-03-30 11:48  cvs

	* bin/pmdistupdate: made sure pmdistupdate needs not arg

2009-03-06 16:43  cvs

	* lib/LEOCHARRE/Dev.pm: made version from finder better

2009-03-06 16:33  cvs

	* bin/pmmakefile: updated pmdistupdate

2009-01-27 10:47  cvs

	* bin/pmmakefile: made pmmakefile include license, author, and
	  optionally die if windows

2009-01-26 15:34  cvs

	* bin/pmmakefile: made it so pmmakefile will insert version into
	  Makefile.PL if it cant find version from

2009-01-23 11:51  cvs

	* bin/pmlicense: added license script for gnu etc

2008-11-05 11:31  cvs

	* bin/pmmanifest.skip: added manifest skip maker

2008-11-05 11:06  cvs

	* bin/pmmakefile: Made sure that the Makefile.PL includes license
	  perl

2008-10-31 10:19  cvs

	* Changes, MANIFEST, Makefile.PL, README, bin/pmassertexe,
	  lib/LEOCHARRE/Dev.pm, lib/LEOCHARRE/PMUsed.pm: Made it so
	  PMUsed.pm does not include things in ./t/ Added README

2008-10-31 09:44  cvs

	* bin/pmmakefile, lib/LEOCHARRE/PMUsed.pm: added a test to see if a
	  module used is in ./t/, to skip it

2008-10-15 14:15  cvs

	* Changes, MANIFEST, bin/pmtodo: made some improvement in pmtodo,
	  still needs real work

2008-10-01 13:48  cvs

	* bin/pmmakefile: made so general basic modules like Cwd and Carp
	  do not require very specific version numbers

2008-10-01 13:38  cvs

	* lib/LEOCHARRE/Dev.pm: added debug and improved where to seek
	  version from in distro

2008-09-22 16:13  cvs

	* MANIFEST, Makefile.PL, bin/pmmanifest: made so pmmanifest will
	  not enter dev entries

2008-09-03 15:37  cvs

	* bin/: pmpod, pmused: made so pmpod works with pm and pl files

2008-09-03 12:51  cvs

	* bin/pmmanifest: made sure to grep out vim swap files

2008-07-28 16:32  cvs

	* MANIFEST, Makefile.PL, bin/pmvercpan: added stuff to auto update
	  perl modules

2008-06-26 07:17  cvs

	* bin/pmchanges: making sure to prune empty changes, only show
	  those with comments

2008-06-25 09:54  cvs

	* MANIFEST, Makefile.PL: added wrapper around cvs2cl

2008-06-25 09:37  cvs

	* Makefile.PL, bin/cvs2cl, bin/pmmakefile, bin/pmused, bin/pmver:
	  added feature to pmmakefile so that you can prune version numbers

2008-06-25 09:27  cvs

	* MANIFEST, Makefile.PL, bin/pmhelp, bin/pmmakefile,
	  bin/pmpodoutline, bin/unbleach: adding automatic changelog script

2008-06-25 09:25  cvs

	* bin/pmver: added options and polished help

2008-06-16 08:52  cvs

	* MANIFEST, Makefile.PL: cleaned up some stuff in distro;

2008-06-08 19:31  cvs

	* MANIFEST, Makefile.PL, bin/pmdistest: added pmdistest

2008-06-08 14:54  cvs

	* bin/pmdistest: planning to make utility to check a archive tar
	  distro

2008-04-15 10:35  cvs

	* bin/pmusing: expanded pmusing to seek sub usage

2008-04-08 22:27  cvs

	* Makefile.PL, bin/pmmakefile, bin/pmmanifest, bin/pmused,
	  lib/LEOCHARRE/Dev.pm, t/0_dev.t: added makefile.pl maker

2008-04-07 09:53  cvs

	* MANIFEST, Makefile.PL, bin/pmpodoutline: added pmpodoutline bin

2007-07-26 18:40  cvs

	* MANIFEST, Makefile.PL, bin/pmtestcheck: added test checked

2007-07-25 12:56  cvs

	* MANIFEST, Makefile.PL, bin/pmused, bin/pmver, bin/unbleach,
	  lib/LEOCHARRE/DEBUG.pm, lib/LEOCHARRE/Dev.pm,
	  lib/LEOCHARRE/PMUsed.pm, t/0.t, t/1_pmused.t, t/inherit_debug.pm,
	  t/mockdev/bin/notperl, t/mockdev/bin/yesperl,
	  t/mockdev/lib/Mock.pm: Initial revision

2007-07-25 12:56  cvs

	* MANIFEST, Makefile.PL, bin/pmused, bin/pmver, bin/unbleach,
	  lib/LEOCHARRE/DEBUG.pm, lib/LEOCHARRE/Dev.pm,
	  lib/LEOCHARRE/PMUsed.pm, t/0.t, t/1_pmused.t, t/inherit_debug.pm,
	  t/mockdev/bin/notperl, t/mockdev/bin/yesperl,
	  t/mockdev/lib/Mock.pm: init



lib/LEOCHARRE/Test.pm
package LEOCHARRE::Test;
use vars qw(@ISA $VERSION @EXPORT);
use strict;
use Exporter;
@ISA = qw/Exporter/;
@EXPORT = qw(ok_part);

my $part =0;
sub ok_part {
   my $label = shift;
   $label ||='';
   
   printf STDERR "\n\n# ------- PART %s %s ------ #\n",$part++, uc($label);
   return 1;
}

1;


lib/LEOCHARRE/Dev.pm
package LEOCHARRE::Dev;
use strict;
use Exporter;
use vars qw($VERSION @ISA @EXPORT_OK %EXPORT_TAGS);
$VERSION = sprintf "%d.%02d", q$Revision: 1.12 $ =~ /(\d+)/g;
@EXPORT_OK = qw(is_pmdist ls_pmdist pmdist_guess_name pmdist_guess_version_from);
@ISA = qw/Exporter/;
%EXPORT_TAGS = ( 'all' => \@EXPORT_OK );

$LEOCHARRE::Dev::DEBUG = 1;
sub debug { print STDERR " - @_\n" if $LEOCHARRE::Dev::DEBUG; 1 }

sub _show_symbol_table {
  require Data::Dumper;
   
  print STDERR " SYMBOL TABLE\n" . Data::Dumper::Dumper(\%LEOCHARRE::Dev::);
  print STDERR "\n";

}

sub is_pmdist {
   my $d = shift;
   require Cwd;
   $d ||= Cwd::cwd();
   
   $d = Cwd::abs_path($d);

   -d $d
      or warn("$d is not dir")
      and return;

   -f "$d/Makefile.PL"
      or warn("$d is not dist, no Makefile.PL")
      and return 0;

   return $d;
}


sub ls_pmdist {
   my $d = shift;
   $d ||= './';
   my $abs_d = is_pmdist($d)  or die;

   my @ls;
   for my $distfile ( sort grep { !/CVS|\.swp$/ } split( /\n/, `find '$abs_d' -type f`) ){
      $distfile=~s/^$abs_d\///;
      push @ls,$distfile;
   }
   defined @ls or warn("nothing in $abs_d ?");
   return @ls;
}


sub pmdist_guess_name {
      my $d = shift;
      $d ||= './';
      
      my $abs_d = is_pmdist($d)  or die;
      $abs_d=~s/^\/.+\/devel\///;
      print STDERR " dist nameguess $abs_d\n";

      $abs_d=~s/.+\///;
      $abs_d=~s/\-/::/g;
      print STDERR " returning $abs_d\n";
   return $abs_d;
   
}

sub pmdist_guess_version_from {
      my $d = shift;
      $d ||= './';
      my $abs_d = is_pmdist($d)  or die;

   if( my $distname = pmdist_guess_name($abs_d) ){
      $distname=~s/::/\//g;
      $distname.='.pm';
      my $pm = $distname;
      if (-f "$abs_d/$distname"){
         debug("found $abs_d/$distname.");

         return $distname;
      }
      elsif( -f "$abs_d/lib/$distname"){
         debug("found $abs_d/lib/$distname.");

         return "lib/$distname";
      }
      debug("Got distname $distname but could not match a file.");
   }

   debug("did not get distname from pmdist_guess_name()"); 
   
   my @pms = grep { /\.pm$/ and !/^blib\// and !/^t\// } ls_pmdist($d);
   @pms and scalar @pms or return;

   my @libpms = grep { /^lib\// } @pms;
   if (@libpms and scalar @libpms ){ # return first
      return $libpms[0];
   }

   # any other?
   return $pms[0];


   #my @pm = grep { /\.pm$/ } ls_pmdist($d) or return;
   #my $first = shift @pm;
   #return $first;

}



1;


__END__

=pod

=head1 NAME

LEOCHARRE::Dev - tools to help manage distro development and maintenance

=head1 DESCRIPTION

This package is a collection of modules and scripts to aid in development.
Most of the functionality is in the set of cli scripts provided in distro,
these reside under bin/

This distro is a set of tools to help me manage a distro- and development.
It is in my namespace because a lot of the scripts and methods here are already available in other packages.

I find these set of tools very valuable.

For example, 

   - automatically generate pod from source
   - analize your distro and tell what modules are deps and what versions are currently installed
   - auto Makefile.PL and MANIFEST generation
   - check out what version of a module is installed and if it's the most current on cpan
   - etc

=head1 SUBS

These subs are contained in the main module. None are exported by default.

=head2 is_pmdist()

Argument is abs path to dist directory.
If no argument provided, uses Cwd.
Returns abs path resolved by Cww::abs_path

   my $abs_dist = is_pmdist();
   my $abs_dist = is_pmdist('./dev/My-Module');

We deem a directory to be a distro if a Makefile.PL file is present therein.

=head2 ls_pmdist()

Argument is path to distro dir.  If left out, tries to use is_pmdist().
Returns list of files relative to dist dir, with no leading slash. (not array ref).

Leaves out CVS entries.

=head1 Executables

A number of useful perl module tools are present to aid in development.

=head1 AUTHOR

Leo Charre leocharre at cpan dot org

=head1 COPYRIGHT

Copyright (c) 2009 Leo Charre. All rights reserved.

=head1 LICENSE

This package is free software; you can redistribute it and/or modify it under the same terms as Perl itself, i.e., under the terms of the "Artistic License" or the "GNU General Public License".

=head1 DISCLAIMER

This package is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

See the "GNU General Public License" for more details.
   
=cut




lib/LEOCHARRE/PMSubs.pm
package LEOCHARRE::PMSubs;
require Exporter;
use vars qw(@ISA $VERSION @EXPORT_OK %EXPORT_TAGS);
use strict;
use Carp;
use warnings;
@ISA = qw(Exporter);
@EXPORT_OK =qw(subs_defined _subs_defined _subs_used subs_used);
%EXPORT_TAGS = ( all => \@EXPORT_OK );
$VERSION = sprintf "%d.%02d", q$Revision: 1.1 $ =~ /(\d+)/g;

sub subs_defined {
   my ($abs_file,$public_only) = @_;
   defined $abs_file or confess('missing file arg');
   $public_only ||=0;

   -f $abs_file or warn("File [$abs_file] is not a file.") and return [];

   my $code = _slurp($abs_file);
   my $subs = _subs_defined($code,$public_only);
   return $subs;
}


sub _subs_defined {
   my ($code,$public_only) = @_;
   defined $code or carp('missing code arg ') and return [];
   $code=~/\w/ or carp('nothing in code arg') and return [];   
   $public_only ||=0;
 
   my $sub={};

   while($code=~/sub\s+([a-zA-Z_]+[\w\: ]*)\s*\{/sg){
      my $_sub = $1;
      $_sub=~s/\s//g;
      next if $public_only and $_sub=~/^_/;
      $sub->{$_sub}++;  
   }   
   
   my @subs = sort keys %$sub;   
   return \@subs;
}



sub _slurp {
   my $abs = shift;
   defined $abs or confess('missing arg');
   
   my $code;
   open(FILE,'<',$abs) or confess($!);;
   while(<FILE>){
      $code.=$_."\n";
   }
   close FILE or confess($!);
   return $code;
}




sub _subs_used {
   my $code = shift;
   defined $code or warn('missing code arg arg') and return {};
   $code=~/\w/ or carp('nothing in code arg') and return {};   
 
   my $sub={};

   while($code=~/(->[a-zA-Z_]+[\w]*|[a-zA-Z_]+[\w]*\()/sg){
      my $_sub = $1; 
      $_sub=~s/^\>|\($//g;
      $sub->{$_sub}++;  
   }  
   
   
   
   return $sub;
}

sub subs_used {
  my ($abs_file) = @_;
  defined $abs_file or confess('missing file arg');

  -f $abs_file or warn("File [$abs_file] is not a file.") and return {};

  my $code = _slurp($abs_file);
  my $sub = _subs_used($code);
  return $sub;
}



1;

__END__

=pod

=head1 NAME

LEOCHARRE::PMSubs - find out what subroutines and or methods are defined in perl code

=head1 SYNOPSIS

   use LEOCHARRE::PMSubs 'subs_defined';
   
   my $codefile = './lib/Module.pm';
   my $subs = subs_defined($codefile);

   map { print STDERR "$codefile : $_\n" } @$subs;   

=head1 DESCRIPTION

This works via regexes and is not perfect, but quick.
This is useful for devel purposes.

=head1 API

=head2 subs_defined()

argument is abs path to perl code file
returns array ref of subs defined in file
optional argument is boolean, if to include only public methods,
methods that do not begin with underscore.
default is 0, all subs/methods.

if no file argument, throws exception
if file does not exist, warns and returns []

=head2 _subs_defined()

argument is code text
returns array ref

optional argument is boolean, if to include only public methods,
methods that do not begin with underscore.

if no code, warns and returns [].


=head2 _subs_used()

argument is code text
for curiosity
returns hash ref with subs used and count of times

CAVEAT: needs to be worked out for kinks

=head2 subs_used()

argument is abs path to code file
returns hash ref with subs used and count of times

=head1 CLI

See bin/pmsubs included in this distro.

=head1 AUTHOR

Leo Charre leocharre at cpan dot org

=head1 SEE ALSO

L<LEOCHARRE::PMUsed>
L<LEOCHARRE::Dev>

=cut


lib/LEOCHARRE/PMUsed.pm
package LEOCHARRE::PMUsed;
use Cwd;
use Carp;
require Exporter;
use vars qw(@ISA $VERSION @EXPORT_OK %EXPORT_TAGS);
use strict;
@ISA = qw(Exporter);
@EXPORT_OK =qw(module_is_installed find_code_files modules_used modules_used_scan_tree);
%EXPORT_TAGS = ( all => \@EXPORT_OK );

$VERSION = sprintf "%d.%02d", q$Revision: 1.6 $ =~ /(\d+)/g;

$LEOCHARRE::PMUsed::DEBUG = 0;
sub DEBUG : lvalue {$LEOCHARRE::PMUsed::DEBUG}
sub debug { my $msg = shift; print STDERR " $msg\n" if DEBUG; }

sub module_is_installed {
   my $module = shift;
   $module or confess('missing argument to module_is_installed');
   no warnings;
   if ( do { eval "require $module;" } ){
      return 1;
   }   
   debug("Missing perl module: $module");
   return 0;
}


sub find_code_files {
   my $abs = shift;
   require File::Find::Rule;

   my $r = new File::Find::Rule();
   $r->file;
   $r->name( qr/\.pl$|\.pm$|\.t$/ );

   
   my @files = $r->in($abs);


  
   #also find in bin, if usr/bin/perl type line is present in first 3 lines

   if (-d "$abs/bin"){
      debug('bin detected');
   
      opendir(DIR, "$abs/bin") or die($!);
      my @bins = grep { -T "$abs/bin/$_" } readdir DIR;
      closedir DIR;

      debug('count in bin '.scalar @bins);

      BINFILES : for (@bins){
         my $abs_bin= "$abs/bin/$_";
         
         my $x =0;
         open(BIN,'<',$abs_bin);         
         LINES : while(<BIN>){
            my $line = $_;
            if ($line=~/#![\/A-Za-z]+bin\/perl\b/){
               push @files, $abs_bin;
               close BIN;         
               next BINFILES;
            }
            if (++$x > 3){
               close BIN;
               next BINFILES;               
            }
         }
         close BIN;      
      }
   }


   
   return \@files;
   
}


sub modules_used {
   my $abs_code = shift;
   (-f $abs_code) or carp("argument to modules_userd() [$abs_code] is not a file.") and return;
   
   my $lines = [];
   my $modules = {};
   my $code;
   open(FILE,'<',$abs_code);
   while(<FILE>){
      push @$lines,$_;
   }
   
   close FILE;

   LINES: for(@$lines){
      my $line = $_;

      $line=~s/#\s.+$//g; # take out comments
      
      if ( $line=~/use base '([^\']+)'/ ){
         $modules->{$1}++;
      }
      elsif ( $line=~/use base qw\W([\s\w\:]+)\W/){
         my @mods = split(/\s/,$1);
         map{ $modules->{$_}++ if $_=~/\w/ } @mods;
      }   

      
      elsif( $line=~/^use\s+([\w\:]+)[\s;]/s ){      
         $modules->{$1}++;            
      }
      elsif ($line=~/^[\W]*use\s+([\w\:]+)[\s;]/s){
         $modules->{$1}++;         
      }
            
      if ( $line=~/require\s+([\w\:]+)\s*;/s ){
         my $module = $1;
         
         if( $module=~/\.pl$|\.pm$/ or $module=~/^\./ ){
            # then skip, it's a include
            next LINES;
         }
         $modules->{$module}++;
      }   
      
   
   }


   
   return $modules;
}


sub modules_used_scan_tree {
   my $abs_dir = shift;
   my $skip_inclusive = shift;
   $skip_inclusive ||= 0;

   my $codefiles = find_code_files($abs_dir) or warn("no code files found?") and return;
 
   # record which they are, so if they are used by others in this palce, we skip,
   # that is, if we have lib/this and lib/that, we cont want to say we need either
   my $skip = {};
   
   for (@$codefiles){   # if we do 'for my $this (@$codefiles) it screws up
      my $path = $_;
      $path=~s/\.pm$// or next;
      $path=~s/^.+lib\///;
      $path=~s/\//\:\:/g;
      debug(" - summed to $path\n");
      $skip->{$path}++;
   }
   
   my $all={};
   
   for(@$codefiles){
      my $modules = modules_used($_) or die;

      MUSED: for my $name (keys %$modules){
         
         #if we have in the tree, skip
         if( $skip_inclusive and exists $skip->{$name}){
            debug("$name is part of distro");
            next MUSED;
         }
         
         my $count = $modules->{$name};

         if(defined $all->{$name}){
            $count = $count + $all->{$name};
         }

         $all->{$name} = $count;         
      }     
   }

   return $all;   
}








1;


__END__

=pod

=head1 NAME

LEOCHARRE::PMUsed - check what modules are used in a directory


=head1 DESCRIPTION

Will recurse all pl, t, pm and any files in bin for use $modulename.
Basically check all perl and find what modules are being used.

Also check if they are installed; optionally.

=head2 module_is_installed()

argument is a module name, like PDF::API2.
returns boolean, warns if not installed.

=head2 find_code_files()

argument is abs path to dir to start from, your dev base etc.
Find all .t .pm .pl files, also if there is a bin dir, will  seek perl scripts
returns array ref with abs paths to files.

=head2 modules_used()

argument is abs path to a perl code file, 
returns hash ref, each key is a module name, the value is the count of times seen.

returns undef and warns if file not there

this list includes modules used and modules required.
if a require such as library.pl is present, this is discarded.
all requires of names with a dot are discarded.

=head2 modules_used_scan_tree()

argument is abs dir, such as /home/me/dev/MySuperThing.
scans entire filetree

returns hash ref, each key is a module name, the value is the count of times seen.


=head1 AUTHOR

Leo Charre leocharre at cpan dot org

=head1 SEE ALSO

LEOCHARRE::Dev



lib/LEOCHARRE/PMInspect.pm
package LEOCHARRE::PMInspect;
use strict;
use warnings;
use Carp;
use Smart::Comments '###';


sub new {
   my($class,$self) = @_;
   $self||={};
   bless $self,$class;
   return $self;
}

sub pm_class {
   my($self,$val) = @_;
   if(defined $val){
      $self->{pm_class} = $val;
   }

   if ((!defined $self->{pm_class} and !defined $self->{pm_instance}) and defined $self->{pm_path}){

         my @pkg = $self->minfo->packages_inside;
         my $class = shift @pkg if @pkg;
         
         $self->{pm_class} = $class;

   }

   
   if( ! defined $self->{pm_class}) {
      if ( my $ref = $self->pm_instance ){
         my $class = ref $ref or die("not ref");
         $self->{pm_class} = $class;
      }
   }
   
   return $self->{pm_class};
}

sub pm_instance {
   my($self,$val) = @_;
   if(defined $val){
      $self->{pm_instance} = $val;
   }

   if( !defined $self->{pm_instance} ){
   
   
      my $class = $self->pm_class or die("pm_instance not passed and pm_class  not set");
      
      
      eval  "use $class;" ;
      my $instance = $class->new() or die;
      $self->{pm_instance} = $instance;
   }
      
   return $self->{pm_instance};

}

sub pm_path {
   my($self,$val) = @_;
   $self->{pm_path} = $val if defined $val;

   unless( defined $self->{pm_path} ){
      $self->minfo;
      $self->{pm_path} =  $self->minfo->file ;   
   }
   return $self->{pm_path};
}






# INSPECTOR OBJECTS

sub _symdump {
   my $self = shift;
   unless( $self->{devel_symdump} ){
      require Devel::Symdump;   
      my $obj = Devel::Symdump->new( $self->pm_class );
      $self->{devel_symdump} = $obj;
   }
   return $self->{devel_symdump};
}





# OUTPUT METHODS

sub dump {
   my $self = shift;
   require Data::Dumper;
   my $string = Data::Dumper::Dumper($self->pm_instance);
   return $string;
}

sub symdump {
   my $self = shift;
   return $self->_symdump->as_string;
}




# ALL OUTPUTS

sub output {
   my $self = shift;

   my $output = sprintf "class: %s, ref: %s\n", $self->pm_class, $self->pm_instance;
   $output   .= sprintf "symdump:\n%s\n", $self->symdump;
   $output   .= sprintf "dump:\n%s\n", $self->dump();
   

   return $output;   
}





sub minfo {
   my $self = shift;

   unless( defined $self->{module_info} ){ 
      require Module::Info;
   
      if ( $self->{pm_path} ){ # if we call via method, will go into endless loop
      
         print STDERR " fromfile \n";
         
         $self->{module_info} = Module::Info->new_from_file( $self->pm_path );
      }

      elsif ( $self->pm_class ){
      
         print STDERR " romclass \n";

         $self->{module_info} = Module::Info->new_from_module( $self->pm_class );         
      }     

      else {
         die("no pm_path or pm_class can be set");
      }
   
   }

   
   return $self->{module_info};
}



1;


__END__









=head1 SYNOPSIS


example 1

   my $i = new LEOCHARRE::PMInspect({
      pm_instance => $instance,
   });


   print $i->output;

example 2

   my $i = new LEOCHARRE::PMInspect({
      pm_class => 'Name::Space',
   });


   print $i->output;

exmaple 3

   my $i = new LEOCHARRE::PMInspect({
      pm_path => './lib/Module.pm',
   });

   print $i->output;


=head2 pm_class()

get name of module

=head2 pm_instance()

retrieve instance

=head2 pm_path()

abs path to file


=head2 _symdump()

returns Devel::Symdump object
no arg

=head2 minfo()

returns Module::Info object
takes no arg

=head2 symdump()

returns symdump output as string

=head2 dump()

returns Data::Dumper output



=head2 output()

returns string for output to STDERR, includes all outputs

=cut


Makefile.PL
use ExtUtils::MakeMaker;

WriteMakefile(
   NAME => 'LEOCHARRE::Dev',
   VERSION_FROM => 'lib/LEOCHARRE/Dev.pm',
   

   # gen /usr/bin/pmused by Fri Mar  6 16:43:53 EST 2009 
   PREREQ_PM => {
      'AppConfig::Std'                    => '1.07',
      'Carp'                              => '1',
      'Cwd'                               => '3',
      'Data::Dumper'                      => '2.121_08',
      'Devel::ModInfo'                    => '0.05',
      'Devel::Symdump'                    => '2.07',
      'Exporter'                          => '5',
      'ExtUtils::MakeMaker'               => '6',
      'File::Basename'                    => '2.74',
      'File::Find'                        => '1.10',
      'File::Find::Rule'                  => '0.30',
      'File::Path'                        => '2.07',
      'Getopt::Std::Strict'               => '1.01',
      'HTTP::Request::Common'             => '1.26',
      'HTTP::Status'                      => '1.28',
      'LEOCHARRE::CLI'                    => '1.18',
      'LEOCHARRE::Class2'                 => '1.16',
      'LEOCHARRE::DEBUG'                  => '1.12',
      'LEOCHARRE::Dir'                    => '1.06',
      'LEOCHARRE::Test'                   => '1.10',
      'LWP::UserAgent'                    => '2.033',
      'Module::Info'                      => '0.3',
      'Net::FTP'                          => '2.75',
      'Smart::Comments'                   => '1',
      'Term::ReadKey'                     => '2.30',
      'Test::Simple'                      => '0',
      'Time::Format'                      => '1.09',
      'WWW::CPAN'                         => '0.011',
   },

   
   LICENSE => 'perl',
   
   
   
   EXE_FILES => [qw(bin/cpanupload
		bin/cvs2cl
		bin/devdebug-httpd
		bin/dyercvsinstall
		bin/pinstall
		bin/pmassertexe
		bin/pmassertos
		bin/pmchanges
		bin/pmcommit
		bin/pmdbhmysql
		bin/pmdistest
		bin/pmdistupdate
		bin/pmfindcode
		bin/pmhelp
		bin/pminspect
		bin/pmlicense
		bin/pmmakefile
		bin/pmmanifest
		bin/pmmanifest.skip
		bin/pmname
		bin/pmpod
		bin/pmpodoutline
		bin/pmreadme
		bin/pmsub
		bin/pmsubs
		bin/pmtestcheck
		bin/pmtestmake
		bin/pmtodo
		bin/pmused
		bin/pmusing
		bin/pmver
		bin/pmvercpan
		bin/pmverstatus
		bin/unbleach)],
);


